Orange/distance/distance.py

```diff
@@ -108,24 +171,43 @@ def compute_distances(self, x1, x2=None):
         - calls a function in Cython that adds the contributions of discrete
           columns
         """
+        def callback(i):
+            return self.callback is not None and self.callback(
+                sum(cb_weights[:cb_counter]) + i *
+                cb_weights[cb_counter] / sum(cb_weights))
+
+        def callback_count():
+            nonlocal cb_counter
+            cb_counter += 1
+            return callback
+
```

Comment:
I don't like two functions sharing a nonlocal variable hosted in the parent function. :) Also, I don't like computing the sum at each step. You can put the sum in a closures:

```python
        def callback_count():
            nonlocal cb_counter
            cb_counter += 1
            base = sum(cb_weights[:cb_counter])
            step = cb_weights[cb_counter] / sum(cb_weights)
            return lambda i: self.callback and self.callback(base + i * step)
```

It doesn't have to be a lambda, it just has to be inside `callback_count`.

Perhaps even bettter:

```python
class StepwiseCallbacks:
    def __init__(self, callback, weights):
        self.step_counter = -1
        self.callback = callback
        self.weights = weights

    def next(self):
        self.step_counter += 1
        base = sum(self.weights[:self.step_counter])
        step = self.weights[self.step_counter] / sum(self.weights)
        return lambda i: self.callback and self.callback(base + i * step)
```

and the in function you have `callbacks = StepwiseCallbacks(self.callback, [20, 10, 50, 5, 15])`, and then create callbacks with `callbacks.next()`.

The solution with the class is better because you can use it in multiple functions.